### PR TITLE
Improve server shutdown behaviour

### PR DESCRIPTION
**Pull request by RagnarW**
Use same shutdown procedure in RaftServer and CatchupServer.

ServerShutdown now handles errors and throws after both channel and
event loop group have been closed. It will also expose possible failure
when shutting down event loop group.

Also fix logging when unable to shutdwon event loop group within given
time

### CODE REVIEW

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                    {
+                        log.error( "Exception when shutting down event loop group", cause );
+                        throw cause;
+                    }
+                } );
+            }
+        }
+    }
+
+    private interface ErrorRunner
```
Comment (martinfurmanski) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: Perhaps just use Runnable?

Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: Runnable will not work since it needs to throw.

Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: We have `ThrowingAction` which does the same but has a type parameter for the throwable type (although it extends Exception)

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupServer.java [hunk 2]
```diff
@@ -208,28 +209,15 @@ private ChannelInboundHandler decoders( CatchupServerProtocol protocol )
...
-        {
-            return;
-        }
-
         log.info( "CatchupServer stopping and unbinding from " + listenAddress );
         try
         {
-            channel.close().sync();
-            channel = null;
+            new ServerShutdown( log, workerGroup, channel ).shutdown();
```
Comment (andrewkerr9000) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupServer.java]: Maybe use a static method, e.g. `ServerShutdown.shutdown( log, workerGroup, channel );`?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java (hunk 1 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: I would have to wrap the cause of the future since it is a throwable. But if this is to be a private sub interface I don't really see the point.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupServer.java (hunk 2 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/CatchupServer.java]: I agree

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        this.log = log;
+        this.eventLoopGroup = eventLoopGroup;
+        this.channel = channel;
+    }
+
+    public void shutdown() throws Throwable
+    {
+        try ( ErrorHandler errorHandler = new ErrorHandler() )
+        {
+            if ( channel != null )
```
Comment (lutovich) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: Why do we need to both close channel and shutdown the event loop group? Isn't it enough to just do the later? Or channel comes from a different group/bootstrap?

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            }
+            catch ( Throwable t )
+            {
+                if ( throwable == null )
+                {
+                    throwable = t;
+                }
+                else
+                {
+                    throwable.addSuppressed( t );
```
Comment (lutovich) [inline enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: Super edge case - this will throw if `t == throwable`

**Review by lutovich — COMMENTED** [review]
Maybe also add couple tests for `ServerShutdown`.

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java (hunk 3 above)
Comment (RagnarW) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: Not sure. I just kept the old logic. It should be enough according to https://netty.io/wiki/user-guide-for-4.x.html#shutting-down-your-application

File: enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java (hunk 1 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/server/ServerShutdown.java]: :+1:
